### ISSUE TITLE
Web: integration test failures in workspace events

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#102365]
Only failing in Web and only running out of sources it seems:

```
 2 failing
  1) vscode API - workspace events: onDidOpenTextDocument, onDidChangeTextDocument, onDidSaveTextDocument:
      AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: /ijnhhxi-NEW <-> /bombyiu
      + expected - actual
      -false
      +true
      at assertEqualPath (extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:284:10)
      at extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:302:30
      at extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:323:36
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:323:18)
      at processTicksAndRejections (internal/process/task_queues.js:89:5)
  2) vscode API - workspace events: onDidSaveTextDocument fires even for non dirty file when saved:
      AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: /eyglyouflp <-> /ytgwtrbtfq
      + expected - actual
      -false
      +true
      at assertEqualPath (extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:284:10)
      at extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:335:30
      at extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:344:32
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts:344:18)
      at processTicksAndRejections (internal/process/task_queues.js:89:5)
```

**Comment by jrieken** [comment]
https://github.com/microsoft/vscode/blob/802572340b1b5ba60d2375375e9a61dcbfeeb6d5/src/vs/workbench/api/browser/mainThreadDocuments.ts#L149

⬆️ that's where the `onDidSaveTextDocument`-event is driven from. So, either a problem with the web-logic or because of the setup. E.g auto-save in web vs manual save in desktop?

**Comment by bpasero** [comment]
@jrieken oh yeah, auto save is on by default in web, good catch! could that be the reason?

**Comment by jrieken** [comment]
idk - I would still expect a save event to be fired when calling save (as done [here](https://github.com/microsoft/vscode/blob/ef9a24f0219ddb7676a5d47ff3d9a0a931464097/extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts#L341)). Tho, it might be the wrong command 🤷

**Comment by bpasero** [comment]
@jrieken this seems to be an issue with how tests are setup. If I only run the workspace tests I see none failing, via `suite.only`:

https://github.com/microsoft/vscode/blob/75e134ae1d398b7d8c971c0cbfdc2e56c319a20a/extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts#L13

A listener not cleaned up?

**Comment by jrieken** [comment]
> A listener not cleaned up?

Only in the web, unlikely. Maybe missing cleanup because of a failure... So, the second could be a false-alarm because the first failed?

**Comment by bpasero** [comment]
This is indeed caused by auto save being enabled by default in web, changing this line fixes the test failure:

https://github.com/microsoft/vscode/blob/c828a5339bca7222fd8f56ab98361a98f736d185/src/vs/workbench/contrib/files/browser/files.contribution.ts#L307

So in theory, a test maybe leaves a model dirty, then continues to run other tests and then after 1s that previous model is being auto saved. 

I see two options:
* we simply disable auto save when tests run
* the owner of these tests checks if possibly a model is not getting disposed at the end of the test and thus has a chance to get auto saved

**Comment by bpasero** [comment]
It would be easier if somehow the name of the test could be part of the models that are being created so that we know from which test the save kicks in.

**Comment by jrieken** [comment]
The weird thing is that the first failing tests already calls [`saveAll`](https://github.com/microsoft/vscode/blob/ef9a24f0219ddb7676a5d47ff3d9a0a931464097/extensions/vscode-api-tests/src/singlefolder-tests/workspace.test.ts#L291). So, either that's not working or another leaked listener modifies files based on those event. The `-NEW` suffix is a hint, there isn't so many tests using that but nothing obvious

**Comment by bpasero** [comment]
I added a new command `_workbench.revertAllDirty` as a replacement for `saveAll` and added it to the two tests that rely on the `onDidSaveTextDocument` event.

Bottom line: `saveAll` only goes over the opened dirty editors and saves them, but does not save dirty working copies that do not have an associated editor. I did not want to go as far as to change `saveAll` to work on hidden working copies so I found this the safer fix.
